ribs/archives/_cvt_archive.py

```diff
@@ -164,16 +170,59 @@ def __init__(self,
         self._chunk_size = chunk_size
 
         if custom_centroids is None:
-            if not isinstance(samples, int):
-                # Validate shape of custom samples. These are ignored when
-                # `custom_centroids` is provided.
-                samples = np.asarray(samples, dtype=self.dtype)
-                if samples.shape[1] != self._measure_dim:
-                    raise ValueError(
-                        f"Samples has shape {samples.shape} but must be of "
-                        f"shape (n_samples, len(ranges)={self._measure_dim})")
-            self._samples = samples
-            self._centroids = None
+            if centroid_method == "kmeans":
+                # Samples are set to custom first and then checked or used.
+                self._samples = samples
+                if not isinstance(samples, int):
+                    # Validate shape of custom samples.
+                    samples = np.asarray(samples, dtype=self.dtype)
+                    if samples.shape[1] != self._measure_dim:
+                        raise ValueError(
+                            f"Samples has shape {samples.shape} but must be of "
+                            f"shape (n_samples, len(ranges)="
+                            f"{self._measure_dim})")
+                else:
+                    self._samples = self._rng.uniform(
+                        self._lower_bounds,
+                        self._upper_bounds,
+                        size=(self._samples, self._measure_dim),
+                    ).astype(self.dtype) if isinstance(self._samples,
+                                                       int) else self._samples
+
+                self._centroids = k_means(self._samples, self._cells,
+                                          **self._k_means_kwargs)[0]
+
+                if self._centroids.shape[0] < self._cells:
+                    raise RuntimeError(
+                        "While generating the CVT, k-means clustering found "
+                        f"{self._centroids.shape[0]} centroids, but this "
+                        f"archive needs {self._cells} cells. This most "
+                        "likely happened because there are too few samples "
+                        "and/or too many cells.")
+
+                if self._use_kd_tree:
```

Comment:
Move this to the end of the init method. The kd tree is separate from the centroid generation as the kd tree is used to find nearest neighbors (make sure you understand this distinction).